src/rgw/rgw_crypt.cc

```diff
@@ -0,0 +1,1338 @@
+    off_t aligned_size = (bl.length() - bl_ofs) & ~(block_size - 1);
+    //save remainder
+    off_t remainder = (bl.length() - bl_ofs) - aligned_size;
+    if(remainder > 0) {
+      bl.copy(bl_ofs + aligned_size, remainder, cache);
+    }
+    if (aligned_size > 0) {
+      bufferlist data;
+      crypt->encrypt(bl, bl_ofs, aligned_size, data, ofs);
+      res=next->handle_data(data, ofs, phandle, pobj, again);
+      ofs += aligned_size;
+      if (res != 0)
+        return res;
+    }
+  }
+  if (bl.length() == 0) {
+    if (cache.length() > 0) {
+      /*flush cached data*/
+      bufferlist data;
+      crypt->encrypt(cache, 0, cache.length(), data, ofs);
+      res = next->handle_data(data, ofs, phandle, pobj, again);
+      ofs += cache.length();
+      cache.clear();
+      if (res != 0)
+        return res;
+    }
+    /*replicate 0-sized handle_data*/
+    res = next->handle_data(cache, ofs, phandle, pobj, again);
+  }
+  return res;
+}
+
+int RGWPutObj_BlockEncrypt::throttle_data(void *handle,
+                                          const rgw_obj& obj,
+                                          bool need_to_wait) {
+  return next->throttle_data(handle, obj, need_to_wait);
+}
+
+std::string create_random_key_selector() {
+  char random[AES_256_KEYSIZE];
+  if (get_random_bytes(&random[0], sizeof(random)) != 0) {
+    dout(0) << "ERROR: cannot get_random_bytes. " << dendl;
+    for (char& v:random) v=rand();
+  }
+  return std::string(random, sizeof(random));
+}
+
+int get_barbican_url(CephContext * const cct,
+                     std::string& url)
+{
+  url = cct->_conf->rgw_barbican_url;
+  if (url.empty()) {
+    ldout(cct, 0) << "ERROR: conf rgw_barbican_url is not set" << dendl;
+    return -EINVAL;
+  }
+
+  if (url.back() != '/') {
+    url.append("/");
+  }
+
+  return 0;
+}
+
+int request_key_from_barbican(CephContext *cct,
+                              boost::string_ref key_id,
+                              boost::string_ref key_selector,
+                              const std::string& barbican_token,
+                              std::string& actual_key) {
+  std::string secret_url;
+  int res;
+  res = get_barbican_url(cct, secret_url);
+  if (res < 0) {
+     return res;
+  }
+  secret_url += "v1/secrets/" + std::string(key_id);
+
+  bufferlist secret_bl;
+  RGWHTTPTransceiver secret_req(cct, &secret_bl);
+  secret_req.append_header("Accept", "application/octet-stream");
+  secret_req.append_header("X-Auth-Token", barbican_token);
+
+  res = secret_req.process("GET", secret_url.c_str());
+  if (res < 0) {
+    return res;
+  }
+  if (secret_req.get_http_status() ==
+      RGWHTTPTransceiver::HTTP_STATUS_UNAUTHORIZED) {
+    return -EACCES;
+  }
+
+  if (secret_req.get_http_status() >=200 &&
+      secret_req.get_http_status() < 300 &&
+      secret_bl.length() == AES_256_KEYSIZE) {
+    actual_key.assign(secret_bl.c_str(), secret_bl.length());
+    memset(secret_bl.c_str(), 0, secret_bl.length());
```

Comment:
nits: indentation of the if/else blocks
```c++
    actual_key = secret_bl.to_str(); // in case secret_bl has more than one buffer
    secret_bl.zero(); // bufferlist has a helper function
```